[WiP] First version of QA acceptance test framework

purbon opened the pull request:
This is the first version, still work in progress, PR to add a complete acceptance test framework to logstash. The code in here is submitted for early review for interested parties, so is still subject to changes when necessary.

The basic functionality of this test are:

* A command executor responsible of sending commands to either a remote machine or to the local environment. This is used during the test execution to process the necessary test steps.
* A collection of rspec helpers that enable a simple and easy description of each test case.
* A collection of public vagrant machines used to run the acceptance test.

To run this you should:

* Cd into the `qa/acceptance` directory.
* Run `vagrant up`  to bootstrap all described machines, you can also bootstrap only an small set of them.
* Cd to the main `qa` directory.
* Run `rake test:ssh_config`: This will generate the current ssh configuration used for the test to locate the machines, look for a file named `.ssh_config` who will have the json information.
* Run `bundle exec rake test:acceptance:all` to run the acceptance test for all described platforms, you can also run  `bundle exec rake test:acceptance:debian` or  `bundle exec rake test:acceptance:centos` to target only one platform.

Hopefully this will report properly without issues.

I would like to have prefered feedback on:

* The rspec helpers workflow and syntax, is this too boilerplate? do you have something missing?
* Some other questions you might have.

This PR content might change, for example around the syntax or the way the rake task description works, all to make it simpler.

Looking forward to get your feedback!

related issue: #5013 #5070

purbon commented:
might be you @ph or @suyograo wanna share some of your thoughts about it?

ph commented:
General comment, can you add the `encoding: utf-8` to all the files.

purbon commented on qa/acceptance/sys/ubuntu/user_bootstrap.sh:
```diff
@@ -0,0 +1,31 @@
+#rake test:install-core
+
+cd
+wget https://download.elastic.co/logstash/logstash/packages/debian/logstash_2.3.1-1_all.deb
```
this is now getting the latest 2.x branch released, however it should get it from local a shared dir as much as possible. The basic idea should be that:

* Pull last released version from URL.
* Push the generated version from local shared directory.

Will make sure to have this implemented next.

What do you think?

ph replied:
What if we keep the scripts the same, they always install a package from the shared directory.

But we add code that is run (could be a rake task) before that to make sure the package is available in the shared directory. So the script could either download the package or ask logstash to build a new one.

purbon replied:
I think we have to keep both, one in the bootstrap script to get the "latest" release and a generated version in the shared directory.

the latest will only be used when performing the upgrade tests.

what do you think?

ph replied:
Yes we need both for the upgrade scenario.

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,66 @@
+      exit(RSpec::Core::Runner.run([Rake::FileList["acceptance/spec/centos/**/*_spec.rb"]]))
+    end
+  end
+end
```
Can we put all the requires a the top? I don't see the need to lazy require here.

ph replied:
I am **suggesting** that we encapsulate the methods that deal with the `ssh_config` file into a class, so we can reuse it or unit test it more easily. 

IMHO  code in the rakefile should be minimal.

purbon replied:
Yes, I agree we should be starting this direction in all our rake tasks. For now code is kind of minimal already, but I'm sure we could simplify it more.

ph replied:
Since its new code, lets do it now, if we don't do it we will never do it ;)

purbon replied:
it's is already done in the current status of the PR 👍

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,66 @@
+    cd "acceptance" do
+      File.write("../.ssh_config", parse(fetch_config[:stdout].split("\n")).to_json)
+    end
+  end
```
Can we give it a better name? With that naming I thought is was a file generated by ssh.
`.vagrant_ssh_config` would make it clear its related to our vagrant setup.

purbon replied:
it's is something generated by ssh, but it contains data to access the vm's over ssh ...., so you know ... naming is hard 💯

purbon replied:
instead of your `.vagrant_ssh_config`, I will be using `.vm_ssh_config` as is less tool dependant and more descriptive as it relates to the vm's used for testing.

ph commented on qa/acceptance/Vagrantfile:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  #ls.vm.network "forwarded_port", guest: 9600, host: 9600
+end
```
Did you consider a hash based configuration for the machine definition like this one?
I find it more concise easier to deal with, but It might be matter of taste.

```ruby
boxes = {
"ubuntu-1204" => { "box" => "elastic/ubuntu-12.04-x86_64" }
"centos-6" => { "box" => "elastic/centos-6-x86_64" }
}

boxes.each |machine_name, config|
machine_type = machine_name.gsub(/[^a-ZA-Z]/,'')
 config.vm.define machine_name do |ls|
     ls.vm.box = config["box"]
   ls.vm.provider "virtualbox" do |v|
     v.memory = 4096
     v.cpus = 4
   end
   ls.vm.synced_folder "../../build", "/logstash-build"
   ls.vm.provision :shell do |sh|
     sh.path = "sys/#{machine_type}/bootstrap.sh"
     sh.privileged = true
  end
 
   ls.vm.provision :shell do |sh|
     sh.path = "sys/{machine_type}/user_bootstrap.sh"
     sh.privileged = false
   end
end
```

I like the idea of letting vagrant decide the port/ips.

ph replied:
I used the name of the distribution for the `machine_type` but it could be a key in the config.

purbon replied:
It for sure makes the configuration shorter for now, my thoughts here are that for now it simplify, as we have nothing platform specific, but if you take a look at https://github.com/elastic/elasticsearch/blob/master/Vagrantfile , the source of this file it gets some thing that are platform specific, so for now I stayed to something like this, but I'm good to make it as DRY as possible.

ph commented on qa/acceptance/spec/centos/lib/install_spec.rb:
```diff
@@ -0,0 +1,31 @@
+  config.servers.each do |host|
+    it_behaves_like "installable", host, config.lookup[host]
+  end
+end
```
I am usually in favor of explicit vs implicit, but I am wondering with all this `host` repetition.
Should we make the `host` implicit on the matcher? Our test will 99% of the time target one machine, the current one.

This would change the test to this format.
```ruby
it "is running in #{name}" do
 start_service("logstash")
 expect("logstash").to be_running
 stop_service("logstash")
end
```

purbon replied:
@ph I agree with you, I find this host usage unnecessary repetitive. I actually spoke about it with @jsvd and the problem we found here is how to easy access to this host variable (param), is not easy to transparently make this option pass to the helpers level in rspec.

do you have an idea?

on the other side, he was ok with this syntax, so I guess we have many opinions here.

ph replied:
Oh yeah, helper scoping in rspec is kinda tricky. But I I though their module got mixed in the generated class. Do they have access to `let` statements?

> he was ok with this syntax, so I guess we have many opinions here.

I find it unecessary, I haven't experimented on this PR but I think we get remove it.

purbon replied:
sure is not necessary, but also not nice to pass this var thought a let statement, I guess the best way here might be pass with metadata, but not sure this is accessible in helpers, will need to investigate.

ph commented on qa/acceptance/spec/centos/spec_helper.rb:
```diff
@@ -0,0 +1,20 @@
+    "centos-6",
+    "centos-7"
+  ]
+end
```
If we use a hash for settings up the vagrant boxes we would be able to reuse the data here.
see my comment on https://github.com/elastic/logstash/pull/5138/files#r60282516

purbon replied:
This point here is only for filtering which boxes belongs to which execution, on next changes we could select here only 1 box.

not sure what we define in the vagrantfile will become later available here.

ph commented on qa/acceptance/spec/config_helper.rb:
```diff
@@ -0,0 +1,19 @@
+require "json"
```
I believe this is a good place to not require anything from the core.

purbon replied:
we, no problem! I stick too much to old behaviour if using plain json parser, 👍 I should remember myself all time!

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,66 @@
+      { :stdout => stdout.read.chomp, :stderr => stderr.read.chomp,
+        :exit_status => wait_thr.value.exitstatus }
+    end
+  end
```
Maybe add a return a message if vagrant is not installed or not available in the current $PATH?

by curiosity, since vagrant is written in ruby, do they have a gem that wrap vagrant? (Not really needed, was a simple curiosity)

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,66 @@
+    end
+    hosts << host
+    hosts
+  end
```
sad 🐼   that vagrant doesn't return a machine readable format

ph commented on qa/acceptance/sys/ubuntu/bootstrap.sh:
```diff
@@ -0,0 +1,22 @@
+BUILD_URL='https://s3-eu-west-1.amazonaws.com/build-eu.elasticsearch.org/logstash'
+URL="$BUILD_URL/$BRANCH/nightly/JDK8/logstash-latest-SNAPSHOT.deb"
+wget $URL
+#dpkg -i "$HOME/logstash-latest-SNAPSHOT.deb"
```
Maybe not needed with the current dependencies, but I think we should future proof it and add this line to make sure apt-get doesn't trigger a dialog.

`export DEBIAN_FRONTEND=noninteractive`

purbon replied:
I understand, but no really sure we need this, this part is more of a leftover from where I pulled this scripts. however I will put this as part of the initial `.sh` script header.

ph commented on qa/acceptance/spec/debian/spec_helper.rb:
```diff
@@ -0,0 +1,20 @@
+
+RSpec.configure do |c|
+  c.include ServiceTester
+end
```
This code is almost the same as https://github.com/elastic/logstash/pull/5138/files#diff-560541d08dfcd2f1980e7a56c37e1b3eR1 we should dry it up.

ph commented on qa/acceptance/sys/centos/user_bootstrap.sh:
```diff
@@ -0,0 +1,31 @@
+#rake bootstrap
+#rake test:install-core
+
+cd
```
Left over?

ph commented on qa/rspec/centos/commands.rb:
```diff
@@ -0,0 +1,55 @@
+      end
+    end
+
+
```
sticky enter key?

ph commented:
I suggest we add a `README.md` in the qa directory and we will populate it as we go with the feature branch, we can add a link to it from the main readme file. WDYT?

ph commented on .gitignore:
```diff
@@ -28,3 +28,8 @@ rspec.xml
+qa/.vagrant
+qa/.rspec
+qa/acceptance/.vagrant
+qa/Gemfile.lock
```
We might want to add `aq/vendor` to this list, in case anyone is doing `bundler install --path vendor` like myself.

purbon replied:
we could, but not sure if makes sense! I know vendor is common directory, but not sure to add it as anyone could use different directories.

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,66 @@
+require 'rspec'
```
Can you add these lines to this file? This will make sure the code actually read the `Gemfile.lock` and the the `.bundler/config`, This is needed if you want to run `bundle install --path vendor`

```ruby
require "rubygems" # Might not be required on all ruby, but doesnt hurt.
require "bundler/setup"
```

ph commented on qa/acceptance/Vagrantfile:
```diff
@@ -0,0 +1,64 @@
+    v.memory = 4096
+    v.cpus = 4
+  end
+ls.vm.synced_folder "../../build", "/logstash-build"
```
I found a small issue when I ran the test, when you start vagrant the `build` folder need to exist before it succesfuly start. Since the `build` is not checked in the repository its possible that the directory wont exist when we try to start the tests.

I suggest for that we try to create it before configuring it.

ph commented on qa/acceptance/sys/centos/user_bootstrap.sh:
```diff
@@ -0,0 +1,31 @@
+#rake test:install-core
+
+cd
+wget https://download.elastic.co/logstash/logstash/packages/centos/logstash-2.3.1-1.noarch.rpm
```
Can we run `wget` in  `–quiet ` mode, because it generates a weird output when provisioning the machine.

ph commented:
I did a a first pass on the code and tried it locally.
I was not able to make it run correctly,

I did the following:
```sh
vagrant up
rake test:ssh_config
rake test:acceptance:debian
```

@purbon @talevy should we also review [runner-tool](https://github.com/purbon/runner-tool) I would be inclined to say yes since we will all depends on it? 

**A few future improvements**

The initial `vagrant up` takes a long time, I am OK with that :)

This Code is completely independent of Logstash and it wont use the bundled jruby to run, are we OK with that strategy? I believe its fine for the first iteration.

We need some kind of `rake test:setup` make sure the system is ready to run test, I had to go through the task and fine what was needed to be run.

purbon commented:
@ph I think I addressed most of your comments, thank you very much for your feedback here, all very valuable!

respect on running the test, there is a small situation I found out, as this is using the `runner-tool` defined in the Gemfile (actually pulled with `:git` option) for rake to find it I somehow need to run `rake test:acceptance:debian` as `bundle exec rake test:acceptance:debian`.

but can not for example run the tasks that interact with vagrant like this because bundler then gets confused with vagrant and the dependencies it needs, if you have any idea how to make this easier I'm very welcome!

purbon commented:
On Tue, Apr 19, 2016 at 9:20 PM Pier-Hugues Pellerin <
notifications@github.com> wrote:

> @purbon <https://github.com/purbon> @talevy <https://github.com/talevy>
> should we also review runner-tool <https://github.com/purbon/runner-tool>
> I would be inclined to say yes since we will all depends on it?
>

sure, the tool is still work in progress, beta, alpha status, but all
comments are for sure very welcome there! collective intelligence is for
sure better than lock down :-P


> *A few future improvements*
>
> The initial vagrant up takes a long time, I am OK with that :)
>
well, we have not much of room to improve this, every time we
bootstrap/provision a new set of VM's this takes time, the idea here is to
have this in mind and use our resources there wisely.


> This Code is completely independent of Logstash and it wont use the
> bundled jruby to run, are we OK with that strategy? I believe its fine for
> the first iteration.
>
The question is why do I want to run this acceptance test with the bundled
ruby, I don't see any benefit for now on it, only problems (as for example
how to redirect it to it, must say our current switch is rather fragile).


> We need some kind of rake test:setup make sure the system is ready to run
> test, I had to go through the task and fine what was needed to be run.
>
Added an initial version of it that basically does vagrant up, but we could
for sure add more detailed tasks.

- purbon


> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/logstash/pull/5138#issuecomment-212075800>
>

ph commented:
> but can not for example run the tasks that interact with vagrant like this because bundler then gets confused with vagrant and the dependencies it needs, if you have any idea how to make this easier I'm very welcome!


Did you try the following `Bundler.with_clean_env`? See: http://bundler.io/v1.3/man/bundle-exec.1.html
```ruby
 def self.execute(cmd)
   Bundler.with_clean_env do
      Open3.popen3(cmd) do |stdin, stdout, stderr, wait_thr|
       { :stdout => stdout.read.chomp, :stderr => stderr.read.chomp,
        :exit_status => wait_thr.value.exitstatus }
      end
   end
end

ph commented:
> The question is why do I want to run this acceptance test with the bundled
ruby, I don't see any benefit for now on it, only problems (as for example
how to redirect it to it, must say our current switch is rather fragile)

Agree its fragile.


> sure, the tool is still work in progress, beta, alpha status, but all
comments are for sure very welcome there! collective intelligence is for
sure better than lock down :-P

To be able to do that, we would need some kind of PR, its is possible for you to create that?
When its ready I think we should move it to the `elastic` organization?

ph commented on qa/Rakefile:
```diff
@@ -0,0 +1,46 @@
+
+  desc "Generate a valid ssh-config"
+  task :ssh_config do
+    require "logstash/json"
```
I do think this will work if you do `bundle exec` since the library will not be available anywhere.
You either need to use the core json or add a dependency on jrjackson.

ph commented on qa/acceptance/Vagrantfile:
```diff
@@ -0,0 +1,60 @@
+    sh.path = "sys/ubuntu/user_bootstrap.sh"
+    sh.privileged = false
+  end
+end
```
What is your opinion on this proposition ? 
https://github.com/elastic/logstash/pull/5138#discussion_r60282516

ph commented:
Added a note concerning the configuration.

purbon commented:
Hi,
 I think I already shared with you, but https://github.com/elastic/logstash/pull/5138#issuecomment-212416758 does not work as it shows this error:

```
Vagrant experienced a version conflict with some installed plugins!\nThis usually happens if you recently upgraded Vagrant. As part of the\nupgrade process, some existing plugins are no longer compatible with\nthis version of Vagrant. The recommended way to fix this is to remove\nyour existing plugins and reinstall them one-by-one. To remove all\nplugins:\n\n    rm -r ~/.vagrant.d/plugins.json ~/.vagrant.d/gems\n\nNote if you have an alternate VAGRANT_HOME environmental variable\nset, the folders above will be in that directory rather than your\nuser's home directory.\n\nThe error message is shown below:\n\nBundler could not find compatible versions for gem \"bundler\":\n  In Gemfile:\n    vagrant (= 1.7.4) ruby depends on\n      bundler (<= 1.10.5, >= 1.5.2) ruby\n\n  Current Bundler version:\n    bundler (1.10.6)\nThis Gemfile requires a different version of Bundler.\nPerhaps you need to update Bundler by running `gem install bundler`?\nCould not find gem 'bundler (<= 1.10.5, >= 1.5.2) ruby', which is required by gem 'vagrant (= 1.7.4) ruby', in any of the sources.
```

I guess this is due to the fact that we would invoke vagrant with the bundler exec context, so it's own dependencies it gets confused. For this time we're good I guess by having this controversy, we can always simplify and fix later on.

purbon commented:
Hi @ph && @talevy , as spoken with @talevy if you have nothing else pending here that you worry about I will move to have a shared branch with this PR as base, we can always iterate from there.

Thanks folks!

ph commented:
@purbon I don't really agree about how the config of the vagrantfile is made but I think it shouldn't block the creation of the branch.

purbon commented:
re vagrant file config, we can always iterate and make it simple, thanks!
will move forward to create the shared branch.

thanks

On Thu, Apr 21, 2016 at 6:40 PM Pier-Hugues Pellerin <
notifications@github.com> wrote:

> @purbon <https://github.com/purbon> I don't really agree about how the
> config of the vagrantfile is made but I think it shouldn't block the
> creation of the branch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/logstash/pull/5138#issuecomment-213005639>
>

purbon commented:
@talevy as discussed, I added a few lines of code in https://github.com/elastic/logstash/pull/5138/commits/e61663e143b3d6db654e18139fb7bdeb2caf6c5e as a first pass on passing hostname information and run only 1 acceptance test, how this helps you.

purbon commented:
squashing and creating the shared branch as agreed here.

purbon commented:
closing this as this got manually upgraded to a shared branch at https://github.com/elastic/logstash/tree/feature/acceptance_test

@ph @talevy looking forward to more code collaboration here! point your PR's there 👍
